### ISSUE TITLE
Circular padding doesn't work when the 'right' side is 0

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#24504]
Reproduction:
```
import torch
import torch.nn.functional as F

inputs = torch.randn(1, 1, 5, 5)
print(F.pad(inputs, [4,0,3,0], mode='circular').shape) # (1,1,5,5) (Wrong)
print(F.pad(inputs, [1,0,0,0], mode='circular').shape) # (1,1,5,5) (Wrong)
print(F.pad(inputs, [1,1,0,0], mode='circular').shape) # (1,1,5,7) (Correct)
print(F.pad(inputs, [2,1,0,0], mode='circular').shape) # (1,1,5,8) (Correct)
```

Correct results:
```
import torch
import torch.nn.functional as F

inputs = torch.randn(1, 1, 5, 5)
print(F.pad(inputs, [4,0,3,0], mode='circular').shape) # (1,1,8,9)
print(F.pad(inputs, [1,0,0,0], mode='circular').shape) # (1,1,5,6)
print(F.pad(inputs, [1,1,0,0], mode='circular').shape) # (1,1,5,7)
print(F.pad(inputs, [2,1,0,0], mode='circular').shape) # (1,1,5,8)
```

cc @SsnL

**Comment by roberthoenig** [comment]
I just hit the same bug.

**Comment by roberthoenig** [comment]
As a workaround, you can define your own 1D circular padding like so:
```python
def padding1d_circular(input, pad):
    return torch.cat([input[:, :, -pad[0]:], input,
                      input[:, :, 0:pad[1]]], dim=2)
```

**Comment by pdlive215** [comment]
The implementation of circular padding is already a bit of a workaround. It's implemented in Python rather than in `torch/csrc`, and it's not defined in `nn.modules.padding`. I have submitted a PR to fix this specific issue, but I think it would make sense going forward to add a proper C++ implementation of circular padding defined in `nn.modules.padding` (and then to remove the workaround Python implementation).

**Comment by mruberry** [comment]
Running this on nightly I get:

```
torch.Size([1, 1, 8, 9])
torch.Size([1, 1, 5, 6])
torch.Size([1, 1, 5, 7])
torch.Size([1, 1, 5, 8])
```

I think this was fixed by https://github.com/pytorch/pytorch/pull/39273 by @cth7.
